src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+        self._alias_cache.clear()
+        if self.close_models:
+            for model in self._model_cache.values():
+                model.close()
+        self._model_cache.clear()
+
+    def resolve(self, url: str) -> str:
+        """Find the canonical URL, following aliases."""
+        original_url = url
+        aliases = 0
+        parts = urllib.parse.urlparse(url)
+        while parts.path.endswith('.alias'):
+            aliases += 1
+            if aliases > MAX_ALIASES:
+                raise models.TooManyAliasesError.with_urls(
+                    f'Reached limit of {MAX_ALIASES} levels of aliases',
+                    url=url, original_url=original_url)
+            if url in self._alias_cache:
+                new_url = self._alias_cache[url]
+            else:
+                with self._session.get(url) as resp:
+                    resp.raise_for_status()
+                    rel_path = resp.text.rstrip()
+                    new_url = urllib.parse.urljoin(resp.url, rel_path)
+                self._alias_cache[url] = new_url
+            _logger.debug('Redirecting from %s to %s', url, new_url)
+            url = new_url
+            parts = urllib.parse.urlparse(url)
+        return url
+
+    def _get_eager(self, url: str) -> Tuple[io.IOBase, str, str]:
+        with self._session.get(url) as resp:
+            resp.raise_for_status()
+            data = resp.content
+            url = resp.url     # Handle HTTP redirects
+        checksum = hashlib.sha256(data).hexdigest()
+        parts = urllib.parse.urlparse(url)
+        match = re.search(r'/sha256_([a-z0-9]+)\.[^/]+$', parts.path)
+        if match:
+            if checksum != match.group(1):
+                raise models.ChecksumError('Content did not match checksum in URL')
+        # typeshed doesn't reflect that BytesIO inherits from BufferedIOBase
+        # (fixed in master, but not in mypy 0.780).
+        return cast(io.IOBase, io.BytesIO(data)), url, checksum
+
+    def _get_lazy(self, url: str) -> Tuple[io.IOBase, str]:
+        fh = HttpFile(self._session, url)
+        return fh, fh.url
+
+    def get(self, url: str, model_class: Type[_M], *,
+            lazy: bool = False) -> _M:
+        """Retrieve a single model.
+
+        The caller must *not* close the retrieved model, as it is cached and
+        a future request for the model would return the closed model. Instead,
+        the caller must close the fetcher once it no longer needs any of the
+        models.
+
+        Parameters
+        ----------
+        lazy
+            If true, create a view of the HDF5 file that only retrieves data
+            as it is needed. Whether this actually allows data to be loaded
+            lazily depends on the `model_class`: some model classes will read
+            all the data into memory on creation, in which case lazy loading
+            may perform significantly worse.
+
+            Lazy loading imposes some additional requirements:
```

Comment:
For the lazy case should we track when completeness of the load. We could then do a checksum (albeit it likely to late, but it would at least help track down errors).

Reply:
The goal of lazy loading is to allow only a subset of the file to be used without paying the cost of downloading and storing the whole thing - if we're able to compute a checksum, the user probably shouldn't have used lazy loading in the first place.

You'd have to allocate memory to hold the whole file, cache each piece as you load it, and keep track of which ranges have been written. That sounds exceptionally complicated.

Reply:
Agreed - it doesn't make much sense, although my point was just that in the case that a user does end up downloading the whole file, even after specifying lazy loading, you could then do a courtesy check.

Reply:
It would have a significant memory usage impact on the common case, and be a lot of complexity, so I'm going to skip it.